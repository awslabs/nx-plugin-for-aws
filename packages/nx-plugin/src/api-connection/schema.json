{
  "$schema": "https://json-schema.org/schema",
  "$id": "APIConnection",
  "title": "",
  "type": "object",
  "properties": {
    "sourceProject": {
      "type": "string",
      "description": "The source project which will call the API",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "Select the source project which will call the API",
      "x-dropdown": "projects"
    },
    "targetProject": {
      "type": "string",
      "description": "The target project containing your API",
      "x-prompt": "Select the target project containing your API",
      "x-dropdown": "projects"
    }
  },
  "required": ["sourceProject", "targetProject"]
}
