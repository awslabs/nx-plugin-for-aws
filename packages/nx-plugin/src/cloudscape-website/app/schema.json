{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    }
  ],
  "type": "object",
  "properties": {
    "name": {
      "description": "The name of the application.",
      "type": "string",
      "$default": {
        "$source": "argv",
        "index": 0
      },
      "x-prompt": "What name would you like to use for the application?",
      "pattern": "^[a-zA-Z][^:]*$"
    },
    "directory": {
      "description": "The directory of the new application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "default": "packages",
      "x-prompt": "What directory would you like to store your application in?"
    }
  },
  "required": ["name"],
  "examplesFile": "../../../docs/application-examples.md"
}
