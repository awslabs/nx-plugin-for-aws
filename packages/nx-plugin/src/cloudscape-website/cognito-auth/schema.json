{
  "$schema": "https://json-schema.org/schema",
  "cli": "nx",
  "$id": "Cloudscape App",
  "title": "Create a Cloudscape React Application",
  "description": "Create a Cloudscape React application for Nx.",
  "examples": [
    {
      "command": "nx g app myapp --directory=myorg",
      "description": "Generate `apps/myorg/myapp` and `apps/myorg/myapp-e2e`"
    },
    {
      "command": "nx g app myapp --classComponent",
      "description": "Use class components instead of functional components"
    }
  ],
  "type": "object",
  "properties": {
    "project": {
      "description": "The root directory of the Cloudscape application.",
      "type": "string",
      "alias": "dir",
      "x-priority": "important",
      "x-prompt": "the root directory for your Cloudscape application",
      "x-dropdown": "projects"
    },
    "cognitoDomain": {
      "description": "The cognito domain prefix to use",
      "type": "string",
      "x-priority": "important",
      "x-prompt": "the cognito domain prefix to use i.e: https://{{cognitoDomain}}-123456789.auth.ap-southeast-2.amazoncognito.com"
    },
    "allowSignup": {
      "description": "Whether to allow self-signup",
      "type": "boolean",
      "default": false,
      "x-prompt": "Enable self signup?",
      "x-priority": "important"
    }
  },
  "required": ["project", "cognitoDomain"]
}
