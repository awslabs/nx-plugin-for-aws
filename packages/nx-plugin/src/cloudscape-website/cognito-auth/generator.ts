/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import {
  joinPathFragments,
  generateFiles,
  Tree,
  readProjectConfiguration,
  addDependenciesToPackageJson,
  installPackagesTask,
  OverwriteStrategy,
} from '@nx/devkit';
import {
  TYPE_DEFINITIONS_DIR,
  PACKAGES_DIR,
  SHARED_CONSTRUCTS_DIR,
  sharedConstructsGenerator,
} from '../../utils/shared-constructs';
import { CognitoAuthGeneratorSchema as CognitoAuthGeneratorSchema } from './schema';
import { runtimeConfigGenerator } from '../runtime-config/generator';
import { tsquery, ast } from '@phenomnomnominal/tsquery';
import {
  ArrowFunction,
  Block,
  factory,
  JsxElement,
  JsxSelfClosingElement,
  NodeFlags,
  SyntaxKind,
  VariableDeclaration,
  InterfaceDeclaration,
  SourceFile,
} from 'typescript';
import { withVersions } from '../../utils/versions';
import {
  addStarExport,
  createJsxElement,
  createJsxElementFromIdentifier,
  destructuredImport,
  replace,
  singleImport,
} from '../../utils/ast';
import { formatFilesInSubtree } from '../../utils/format';
export async function cognitoAuthGenerator(
  tree: Tree,
  options: CognitoAuthGeneratorSchema,
) {
  const srcRoot = readProjectConfiguration(tree, options.project).sourceRoot;
  if (
    tree.exists(joinPathFragments(srcRoot, 'components/CognitoAuth/index.tsx'))
  ) {
    throw new Error(
      `This generator has already been run on ${options.project}.`,
    );
  }
  await runtimeConfigGenerator(tree, {
    project: options.project,
  });
  await sharedConstructsGenerator(tree);
  // Add ICognitoProps interface and update IRuntimeConfig
  const runtimeConfigPath = joinPathFragments(
    PACKAGES_DIR,
    TYPE_DEFINITIONS_DIR,
    'src',
    'runtime-config.ts',
  );
  const runtimeConfigContent = tree.read(runtimeConfigPath).toString();
  const sourceFile = ast(runtimeConfigContent);
  // Check if ICognitoProps interface exists
  const existingCognitoProps = tsquery.query(
    sourceFile,
    'InterfaceDeclaration[name.text="ICognitoProps"]',
  );
  // Check if cognitoProps property exists in IRuntimeConfig
  const existingCognitoPropsInConfig = tsquery.query(
    sourceFile,
    'InterfaceDeclaration[name.text="IRuntimeConfig"] PropertySignature[name.text="cognitoProps"]',
  );
  let updatedContent = sourceFile;
  // Add ICognitoProps interface if it doesn't exist
  if (existingCognitoProps.length === 0) {
    const cognitoPropsInterface = factory.createInterfaceDeclaration(
      [factory.createModifier(SyntaxKind.ExportKeyword)],
      factory.createIdentifier('ICognitoProps'),
      undefined,
      undefined,
      [
        factory.createPropertySignature(
          undefined,
          factory.createIdentifier('region'),
          undefined,
          factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        ),
        factory.createPropertySignature(
          undefined,
          factory.createIdentifier('identityPoolId'),
          undefined,
          factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        ),
        factory.createPropertySignature(
          undefined,
          factory.createIdentifier('userPoolId'),
          undefined,
          factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        ),
        factory.createPropertySignature(
          undefined,
          factory.createIdentifier('userPoolWebClientId'),
          undefined,
          factory.createKeywordTypeNode(SyntaxKind.StringKeyword),
        ),
      ],
    );
    updatedContent = tsquery.map(
      updatedContent,
      'SourceFile',
      (node: SourceFile) => {
        return factory.updateSourceFile(node, [
          cognitoPropsInterface,
          ...node.statements,
        ]);
      },
    );
  }
  // Add cognitoProps to IRuntimeConfig if it doesn't exist
  if (existingCognitoPropsInConfig.length === 0) {
    updatedContent = tsquery.map(
      updatedContent,
      'InterfaceDeclaration[name.text="IRuntimeConfig"]',
      (node: InterfaceDeclaration) => {
        return factory.updateInterfaceDeclaration(
          node,
          node.modifiers,
          node.name,
          node.typeParameters,
          node.heritageClauses,
          [
            ...node.members,
            factory.createPropertySignature(
              undefined,
              factory.createIdentifier('cognitoProps'),
              undefined,
              factory.createTypeReferenceNode('ICognitoProps', undefined),
            ),
          ],
        );
      },
    );
  }
  // Only write if changes were made
  if (updatedContent !== sourceFile) {
    tree.write(runtimeConfigPath, updatedContent.getFullText());
  }
  const identityPath = joinPathFragments(
    PACKAGES_DIR,
    SHARED_CONSTRUCTS_DIR,
    'src',
    'core',
    'user-identity.ts',
  );
  generateFiles(
    tree,
    joinPathFragments(__dirname, 'files', 'app'),
    srcRoot,
    options,
    {
      overwriteStrategy: OverwriteStrategy.KeepExisting,
    },
  );
  if (!tree.exists(identityPath)) {
    generateFiles(
      tree,
      joinPathFragments(__dirname, 'files', SHARED_CONSTRUCTS_DIR),
      joinPathFragments(PACKAGES_DIR, SHARED_CONSTRUCTS_DIR),
      {
        allowSignup: options.allowSignup,
        cognitoDomain: options.cognitoDomain,
      },
      {
        overwriteStrategy: OverwriteStrategy.KeepExisting,
      },
    );
    addDependenciesToPackageJson(
      tree,
      withVersions([
        'oidc-client-ts',
        'react-oidc-context',
        '@aws-cdk/aws-cognito-identitypool-alpha',
      ]),
      {},
    );
    addStarExport(
      tree,
      joinPathFragments(
        PACKAGES_DIR,
        SHARED_CONSTRUCTS_DIR,
        'src',
        'core',
        'index.ts',
      ),
      './user-identity.js',
    );
  }
  const mainTsxPath = joinPathFragments(srcRoot, 'main.tsx');
  singleImport(tree, mainTsxPath, 'CognitoAuth', './components/CognitoAuth');
  replace(
    tree,
    mainTsxPath,
    'JsxElement[openingElement.tagName.name="RuntimeConfigProvider"]',
    (node: JsxElement) =>
      createJsxElement(
        node.openingElement,
        [createJsxElementFromIdentifier('CognitoAuth', node.children)],
        node.closingElement,
      ),
  );
  // Update App Layout
  const appLayoutTsxPath = joinPathFragments(
    srcRoot,
    'components',
    'AppLayout',
    'index.tsx',
  );
  if (tree.exists(appLayoutTsxPath)) {
    const contents = tree.read(appLayoutTsxPath).toString();
    let updatedContents = destructuredImport(
      tree,
      appLayoutTsxPath,
      ['useAuth'],
      'react-oidc-context',
    );
    updatedContents = tsquery
      .map(
        ast(updatedContents),
        'VariableDeclaration',
        (node: VariableDeclaration) => {
          // Only process if this is the App component
          if (node.name.getText() !== 'AppLayout') {
            return node;
          }
          const arrowFunction = node.initializer as ArrowFunction;
          const functionBody = arrowFunction.body as Block;
          // Create our new declaration
          const authDeclaration = factory.createVariableStatement(
            undefined,
            factory.createVariableDeclarationList(
              [
                factory.createVariableDeclaration(
                  factory.createObjectBindingPattern([
                    factory.createBindingElement(
                      undefined,
                      undefined,
                      factory.createIdentifier('user'),
                      undefined,
                    ),
                    factory.createBindingElement(
                      undefined,
                      undefined,
                      factory.createIdentifier('removeUser'),
                      undefined,
                    ),
                    factory.createBindingElement(
                      undefined,
                      undefined,
                      factory.createIdentifier('signoutRedirect'),
                      undefined,
                    ),
                    factory.createBindingElement(
                      undefined,
                      undefined,
                      factory.createIdentifier('clearStaleState'),
                      undefined,
                    ),
                  ]),
                  undefined,
                  undefined,
                  factory.createCallExpression(
                    factory.createIdentifier('useAuth'),
                    undefined,
                    [],
                  ),
                ),
              ],
              NodeFlags.Const,
            ),
          );
          // Add as first statement
          const newStatements = [authDeclaration, ...functionBody.statements];
          // Create new arrow function with updated body
          const newArrowFunction = factory.updateArrowFunction(
            arrowFunction,
            arrowFunction.modifiers,
            arrowFunction.typeParameters,
            arrowFunction.parameters,
            arrowFunction.type,
            arrowFunction.equalsGreaterThanToken,
            factory.createBlock(newStatements, true),
          );
          // Update the variable declaration
          return factory.updateVariableDeclaration(
            node,
            node.name,
            node.exclamationToken,
            node.type,
            newArrowFunction,
          );
        },
      )
      .getFullText();
    // TODO: update utils if they exist by appending to the array
    updatedContents = tsquery
      .map(
        ast(updatedContents),
        'JsxSelfClosingElement[tagName.text="TopNavigation"]',
        (node: JsxSelfClosingElement) => {
          // Create the utilities attribute
          const utilitiesAttribute = factory.createJsxAttribute(
            factory.createIdentifier('utilities'),
            factory.createJsxExpression(
              undefined,
              factory.createArrayLiteralExpression(
                [
                  factory.createObjectLiteralExpression([
                    factory.createPropertyAssignment(
                      'type',
                      factory.createStringLiteral('menu-dropdown'),
                    ),
                    factory.createPropertyAssignment(
                      'text',
                      factory.createTemplateExpression(
                        factory.createTemplateHead(''),
                        [
                          factory.createTemplateSpan(
                            factory.createElementAccessChain(
                              factory.createPropertyAccessChain(
                                factory.createIdentifier('user'),
                                factory.createToken(
                                  SyntaxKind.QuestionDotToken,
                                ),
                                factory.createIdentifier('profile'),
                              ),
                              factory.createToken(SyntaxKind.QuestionDotToken),
                              factory.createStringLiteral('cognito:username'),
                            ),
                            factory.createTemplateTail(''),
                          ),
                        ],
                      ),
                    ),
                    factory.createPropertyAssignment(
                      'iconName',
                      factory.createStringLiteral('user-profile-active'),
                    ),
                    factory.createPropertyAssignment(
                      'onItemClick',
                      factory.createArrowFunction(
                        undefined,
                        undefined,
                        [
                          factory.createParameterDeclaration(
                            undefined,
                            undefined,
                            factory.createIdentifier('e'),
                            undefined,
                            undefined,
                            undefined,
                          ),
                        ],
                        undefined,
                        factory.createToken(SyntaxKind.EqualsGreaterThanToken),
                        factory.createBlock(
                          [
                            factory.createIfStatement(
                              factory.createBinaryExpression(
                                factory.createPropertyAccessExpression(
                                  factory.createPropertyAccessExpression(
                                    factory.createIdentifier('e'),
                                    factory.createIdentifier('detail'),
                                  ),
                                  factory.createIdentifier('id'),
                                ),
                                factory.createToken(
                                  SyntaxKind.EqualsEqualsEqualsToken,
                                ),
                                factory.createStringLiteral('signout'),
                              ),
                              factory.createBlock(
                                [
                                  factory.createExpressionStatement(
                                    factory.createCallExpression(
                                      factory.createIdentifier('removeUser'),
                                      undefined,
                                      [],
                                    ),
                                  ),
                                  factory.createExpressionStatement(
                                    factory.createCallExpression(
                                      factory.createIdentifier(
                                        'signoutRedirect',
                                      ),
                                      undefined,
                                      [
                                        factory.createObjectLiteralExpression([
                                          factory.createPropertyAssignment(
                                            'post_logout_redirect_uri',
                                            factory.createPropertyAccessExpression(
                                              factory.createPropertyAccessExpression(
                                                factory.createIdentifier(
                                                  'window',
                                                ),
                                                factory.createIdentifier(
                                                  'location',
                                                ),
                                              ),
                                              factory.createIdentifier(
                                                'origin',
                                              ),
                                            ),
                                          ),
                                          factory.createPropertyAssignment(
                                            'extraQueryParams',
                                            factory.createObjectLiteralExpression(
                                              [
                                                factory.createPropertyAssignment(
                                                  'redirect_uri',
                                                  factory.createPropertyAccessExpression(
                                                    factory.createPropertyAccessExpression(
                                                      factory.createIdentifier(
                                                        'window',
                                                      ),
                                                      factory.createIdentifier(
                                                        'location',
                                                      ),
                                                    ),
                                                    factory.createIdentifier(
                                                      'origin',
                                                    ),
                                                  ),
                                                ),
                                                factory.createPropertyAssignment(
                                                  'response_type',
                                                  factory.createStringLiteral(
                                                    'code',
                                                  ),
                                                ),
                                              ],
                                            ),
                                          ),
                                        ]),
                                      ],
                                    ),
                                  ),
                                  factory.createExpressionStatement(
                                    factory.createCallExpression(
                                      factory.createIdentifier(
                                        'clearStaleState',
                                      ),
                                      undefined,
                                      [],
                                    ),
                                  ),
                                ],
                                true,
                              ),
                            ),
                          ],
                          true,
                        ),
                      ),
                    ),
                    factory.createPropertyAssignment(
                      'items',
                      factory.createArrayLiteralExpression([
                        factory.createObjectLiteralExpression([
                          factory.createPropertyAssignment(
                            'id',
                            factory.createStringLiteral('signout'),
                          ),
                          factory.createPropertyAssignment(
                            'text',
                            factory.createStringLiteral('Sign out'),
                          ),
                        ]),
                      ]),
                    ),
                  ]),
                ],
                true,
              ),
            ),
          );
          // Add the utilities attribute to existing attributes
          return factory.createJsxSelfClosingElement(
            node.tagName,
            node.typeArguments,
            factory.createJsxAttributes([
              ...node.attributes.properties,
              utilitiesAttribute,
            ]),
          );
        },
      )
      .getFullText();
    if (contents !== updatedContents) {
      tree.write(appLayoutTsxPath, updatedContents);
    }
  } else {
    console.info(
      `Skipping update to ${appLayoutTsxPath} as it does not exist.`,
    );
  }
  // End update App Layout
  await formatFilesInSubtree(tree);
  return () => {
    installPackagesTask(tree);
  };
}
export default cognitoAuthGenerator;
