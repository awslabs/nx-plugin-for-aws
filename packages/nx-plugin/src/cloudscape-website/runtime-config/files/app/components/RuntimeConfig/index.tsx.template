/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import ErrorMessage from '@aws-northstar/ui/components/CognitoAuth/components/ErrorMessage';
import { Spinner } from '@cloudscape-design/components';
import React, { createContext, useEffect, useState } from 'react';
import type { IRuntimeConfig } from '<%= scopeAlias %>common-types';

/**
 * Context for storing the runtimeConfig.
 */
export const RuntimeConfigContext = createContext<IRuntimeConfig | undefined>(
  undefined
);

/**
 * Sets up the runtimeConfig.
 *
 * This assumes a runtime-config.json file is present at '/'.
 */
const RuntimeConfigProvider: React.FC<any> = ({ children }) => {
  const [runtimeConfig, setRuntimeConfig] = useState<
    IRuntimeConfig | undefined
  >();
  const [error, setError] = useState<string | undefined>();

  useEffect(() => {
    fetch('/runtime-config.json')
      .then((response) => {
        return response.json();
      })
      .then((_runtimeConfig) => {
        setRuntimeConfig(_runtimeConfig as IRuntimeConfig);
      })
      .catch(() => {
        setError('No runtime-config.json detected');
      });
  }, [setRuntimeConfig]);

  return error ? (
    <ErrorMessage>{error}</ErrorMessage>
  ) : runtimeConfig ? (
    <RuntimeConfigContext.Provider value={runtimeConfig}>
      {children}
    </RuntimeConfigContext.Provider>
  ) : (
    <Spinner />
  );
};

export default RuntimeConfigProvider;
