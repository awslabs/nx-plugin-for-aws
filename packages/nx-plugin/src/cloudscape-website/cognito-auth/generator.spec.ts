/**
 * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 * SPDX-License-Identifier: Apache-2.0
 */
import { Tree } from '@nx/devkit';
import { cognitoAuthGenerator } from './generator';
import { CognitoAuthGeneratorSchema } from './schema';
import { createTreeUsingTsSolutionSetup } from '../../utils/test';
describe('cognito-auth generator', () => {
  let tree: Tree;
  const options: CognitoAuthGeneratorSchema = {
    project: 'test-project',
    cognitoDomain: 'test',
    allowSignup: true,
  };
  beforeEach(() => {
    tree = createTreeUsingTsSolutionSetup();
    // Set up a mock project structure
    tree.write(
      'packages/test-project/project.json',
      JSON.stringify({
        name: 'test-project',
        sourceRoot: 'packages/test-project/src',
      }),
    );
  });
  it('should generate files', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { BrowserRouter } from 'react-router-dom';

      export function App() {
        return (
          <RuntimeConfigProvider>
            <BrowserRouter>Hello World</BrowserRouter>
          </RuntimeConfigProvider>
        );
      }
    `,
    );
    await cognitoAuthGenerator(tree, options);
    // Verify component files are generated
    expect(
      tree.exists('packages/test-project/src/components/CognitoAuth/index.tsx'),
    ).toBeTruthy();
    // Verify shared constructs files are generated
    expect(
      tree.exists('packages/common/constructs/src/core/user-identity.ts'),
    ).toBeTruthy();
    expect(
      tree.exists('packages/common/constructs/src/core/index.ts'),
    ).toBeTruthy();
    // Create snapshot of the generated files
    expect(
      tree
        .read('packages/test-project/src/components/CognitoAuth/index.tsx')
        .toString(),
    ).toMatchSnapshot('cognito-auth-component');
    expect(
      tree.read('packages/common/constructs/src/core/index.ts').toString(),
    ).toMatchSnapshot('identity-index');
    expect(
      tree
        .read('packages/common/constructs/src/core/user-identity.ts')
        .toString(),
    ).toMatchSnapshot('user-identity');
  });
  it('should update main.tsx when RuntimeConfigProvider exists', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      import { RuntimeConfigProvider } from './components/RuntimeConfig';
      import { BrowserRouter } from 'react-router-dom';

      export function App() {
        return (
          <RuntimeConfigProvider>
            <BrowserRouter>Hello World</BrowserRouter>
          </RuntimeConfigProvider>
        );
      }
    `,
    );
    await cognitoAuthGenerator(tree, options);
    const mainTsxContent = tree
      .read('packages/test-project/src/main.tsx')
      .toString();
    // Verify CognitoAuth import is added
    expect(mainTsxContent).toContain(
      "import CognitoAuth from './components/CognitoAuth'",
    );
    // Verify CognitoAuth component is wrapped around children
    expect(mainTsxContent).toContain('<CognitoAuth>');
    expect(mainTsxContent).toContain('</CognitoAuth>');
    // Create snapshot of the modified main.tsx
    expect(mainTsxContent).toMatchSnapshot('main-tsx-with-runtime-config');
  });
  it('should handle main.tsx without RuntimeConfigProvider', async () => {
    // Setup main.tsx without RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
      export function App() {
        return (
          <div>Hello World</div>
        );
      }
    `,
    );
    await expect(
      async () => await cognitoAuthGenerator(tree, options),
    ).rejects.toThrowError();
  });
  it('should handle missing main.tsx', async () => {
    await expect(
      async () => await cognitoAuthGenerator(tree, options),
    ).rejects.toThrowError();
  });
  it('should update shared constructs index.ts', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
        import { RuntimeConfigProvider } from './components/RuntimeConfig';
        import { BrowserRouter } from 'react-router-dom';

        export function App() {
          return (
            <RuntimeConfigProvider>
              <BrowserRouter>Hello World</BrowserRouter>
            </RuntimeConfigProvider>
          );
        }
      `,
    );
    // Setup initial shared constructs index
    tree.write(
      'packages/common/constructs/src/index.ts',
      'export const dummy = true;',
    );
    await cognitoAuthGenerator(tree, options);
    const indexContent = tree
      .read('packages/common/constructs/src/core/index.ts')
      .toString();
    // Verify identity export is added
    expect(indexContent).toContain("export * from './user-identity.js'");
    // Create snapshot of the modified index
    expect(indexContent).toMatchSnapshot('common/constructs-index');
  });
  it('should add required dependencies', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
        import { RuntimeConfigProvider } from './components/RuntimeConfig';
        import { BrowserRouter } from 'react-router-dom';

        export function App() {
          return (
            <RuntimeConfigProvider>
              <BrowserRouter>Hello World</BrowserRouter>
            </RuntimeConfigProvider>
          );
        }
      `,
    );
    await cognitoAuthGenerator(tree, options);
    // Read package.json
    const packageJson = JSON.parse(tree.read('package.json').toString());
    // Verify dependencies are added
    expect(packageJson.dependencies).toMatchObject({
      constructs: expect.any(String),
      'aws-cdk-lib': expect.any(String),
      '@aws-cdk/aws-cognito-identitypool-alpha': expect.any(String),
      'oidc-client-ts': expect.any(String),
      'react-oidc-context': expect.any(String),
    });
  });
  it('should not be able to run the generator multiple times', async () => {
    // Setup main.tsx with RuntimeConfigProvider
    tree.write(
      'packages/test-project/src/main.tsx',
      `
        import { RuntimeConfigProvider } from './components/RuntimeConfig';
        import { BrowserRouter } from 'react-router-dom';

        export function App() {
          return (
            <RuntimeConfigProvider>
              <BrowserRouter>Hello World</BrowserRouter>
            </RuntimeConfigProvider>
          );
        }
      `,
    );
    // First run to create files
    await cognitoAuthGenerator(tree, options);
    // Run generator again
    await expect(
      async () => await cognitoAuthGenerator(tree, options),
    ).rejects.toThrowErrorMatchingSnapshot();
  });
});
