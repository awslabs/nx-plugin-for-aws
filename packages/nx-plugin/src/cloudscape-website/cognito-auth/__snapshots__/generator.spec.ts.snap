// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`cognito-auth generator > should generate files > cognito-auth-component 1`] = `
"import React, { PropsWithChildren, useEffect } from 'react';
import { AuthProvider, AuthProviderProps, useAuth } from 'react-oidc-context';
import { Alert, Spinner } from '@cloudscape-design/components';
import { useRuntimeConfig } from '../../hooks/useRuntimeConfig';

/**
 * Sets up the Cognito auth.
 *
 * This assumes a runtime-config.json file is present at '/'. In order for Auth to be set up automatically,
 * the runtime-config.json must have the cognitoProps set.
 */
const CognitoAuth: React.FC<PropsWithChildren> = ({ children }) => {
  const { cognitoProps } = useRuntimeConfig();

  if (!cognitoProps) {
    return (
      <Alert type="error" header="Runtime config configuration error">
        The cognitoProps have not been configured in the runtime-config.json.
      </Alert>
    );
  }

  const cognitoAuthConfig: AuthProviderProps = {
    authority: \`https://cognito-idp.\${cognitoProps.region}.amazonaws.com/\${cognitoProps.userPoolId}\`,
    client_id: cognitoProps.userPoolWebClientId,
    redirect_uri: window.location.origin,
    response_type: 'code',
    scope: 'email openid profile',
  };

  return (
    <AuthProvider {...cognitoAuthConfig}>
      <CognitoAuthInternal>{children}</CognitoAuthInternal>
    </AuthProvider>
  );
};

const CognitoAuthInternal: React.FC<PropsWithChildren> = ({ children }) => {
  const auth = useAuth();

  useEffect(() => {
    if (!auth.isAuthenticated && !auth.isLoading) {
      auth.signinRedirect();
    }
  }, [auth]);

  if (auth.isAuthenticated) {
    return children;
  } else if (auth.error) {
    return (
      <Alert type="error" header="Configuration error">
        Error contacting Cognito. Please check your runtime-config.json is
        configured with the correct endpoints.
      </Alert>
    );
  } else {
    return <Spinner />;
  }
};

export default CognitoAuth;
"
`;

exports[`cognito-auth generator > should generate files > identity-index 1`] = `
"export * from './user-identity.js';
export * from './runtime-config.js';
"
`;

exports[`cognito-auth generator > should generate files > user-identity 1`] = `
"import {
  IdentityPool,
  UserPoolAuthenticationProvider,
} from '@aws-cdk/aws-cognito-identitypool-alpha';
import { CfnOutput, Duration, Lazy, Stack } from 'aws-cdk-lib';
import {
  AccountRecovery,
  CfnManagedLoginBranding,
  CfnUserPoolDomain,
  FeaturePlan,
  Mfa,
  OAuthScope,
  UserPool,
  UserPoolClient,
} from 'aws-cdk-lib/aws-cognito';
import { Construct } from 'constructs';
import { RuntimeConfig } from './runtime-config.js';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';

const WEB_CLIENT_ID = 'WebClient';
/**
 * Creates a UserPool and Identity Pool with sane defaults configured intended for usage from a web client.
 */
export class UserIdentity extends Construct {
  public readonly region: string;
  public readonly identityPool: IdentityPool;
  public readonly userPool: UserPool;
  public readonly userPoolClient: UserPoolClient;
  public readonly userPoolDomain: CfnUserPoolDomain;

  constructor(scope: Construct, id: string) {
    super(scope, id);

    this.region = Stack.of(this).region;
    this.userPool = this.createUserPool();
    this.userPoolDomain = this.createUserPoolDomain(this.userPool);
    this.userPoolClient = this.createUserPoolClient(this.userPool);
    this.identityPool = this.createIdentityPool(
      this.userPool,
      this.userPoolClient,
    );
    this.createManagedLoginBranding(
      this.userPool,
      this.userPoolClient,
      this.userPoolDomain,
    );

    RuntimeConfig.ensure(this).config.cognitoProps = {
      region: Stack.of(this).region,
      identityPoolId: this.identityPool.identityPoolId,
      userPoolId: this.userPool.userPoolId,
      userPoolWebClientId: this.userPoolClient.userPoolClientId,
    };

    new CfnOutput(this, \`\${id}-UserPoolId\`, {
      value: this.userPool.userPoolId,
    });

    new CfnOutput(this, \`\${id}-IdentityPoolId\`, {
      value: this.identityPool.identityPoolId,
    });
  }

  private createUserPool = () =>
    new UserPool(this, 'UserPool', {
      deletionProtection: true,
      passwordPolicy: {
        minLength: 8,
        requireLowercase: true,
        requireUppercase: true,
        requireDigits: true,
        requireSymbols: true,
        tempPasswordValidity: Duration.days(3),
      },
      mfa: Mfa.REQUIRED,
      featurePlan: FeaturePlan.ESSENTIALS,
      mfaSecondFactor: { sms: true, otp: true },
      signInCaseSensitive: false,
      signInAliases: { username: true, email: true },
      accountRecovery: AccountRecovery.EMAIL_ONLY,
      selfSignUpEnabled: true,
      standardAttributes: {
        phoneNumber: { required: false },
        email: { required: true },
        givenName: { required: true },
        familyName: { required: true },
      },
      autoVerify: {
        email: true,
        phone: true,
      },
      keepOriginal: {
        email: true,
        phone: true,
      },
    });

  private createUserPoolDomain = (userPool: UserPool) =>
    new CfnUserPoolDomain(this, 'UserPoolDomain', {
      domain: \`test-\${Stack.of(this).account}\`,
      userPoolId: userPool.userPoolId,
      managedLoginVersion: 2,
    });

  private createUserPoolClient = (userPool: UserPool) => {
    const lazilyComputedCallbackUrls = Lazy.list({
      produce: () =>
        [
          'http://localhost:4200',
          \`https://\${Stack.of(this).region}.console.aws.amazon.com\`,
        ].concat(
          this.findCloudFrontDistributions().map(
            (d) => \`https://\${d.domainName}\`,
          ),
        ),
    });

    return userPool.addClient(WEB_CLIENT_ID, {
      authFlows: {
        userPassword: true,
        userSrp: true,
        user: true,
      },
      oAuth: {
        flows: {
          authorizationCodeGrant: true,
        },
        scopes: [OAuthScope.EMAIL, OAuthScope.OPENID, OAuthScope.PROFILE],
        callbackUrls: lazilyComputedCallbackUrls,
        logoutUrls: lazilyComputedCallbackUrls,
      },
      preventUserExistenceErrors: true,
    });
  };

  private createIdentityPool = (
    userPool: UserPool,
    userPoolClient: UserPoolClient,
  ) => {
    const identityPool = new IdentityPool(this, 'IdentityPool');

    identityPool.addUserPoolAuthentication(
      new UserPoolAuthenticationProvider({
        userPool,
        userPoolClient,
      }),
    );

    return identityPool;
  };

  private createManagedLoginBranding = (
    userPool: UserPool,
    userPoolClient: UserPoolClient,
    userPoolDomain: CfnUserPoolDomain,
  ) => {
    new CfnManagedLoginBranding(this, 'ManagedLoginBranding', {
      userPoolId: userPool.userPoolId,
      clientId: userPoolClient.userPoolClientId,
      useCognitoProvidedValues: true,
    }).node.addDependency(userPoolClient, userPool, userPoolDomain);
  };

  private findCloudFrontDistributions = (): Distribution[] =>
    Stack.of(this)
      .node.findAll()
      .filter((child) => child instanceof Distribution);
}
"
`;

exports[`cognito-auth generator > should not be able to run the generator multiple times 1`] = `[Error: This generator has already been run on test-project.]`;

exports[`cognito-auth generator > should update main.tsx when RuntimeConfigProvider exists > main-tsx-with-runtime-config 1`] = `
"import CognitoAuth from './components/CognitoAuth';
import RuntimeConfigProvider from './components/RuntimeConfig';
import { RuntimeConfigProvider } from './components/RuntimeConfig';
import { RouterProvider, createRouter } from '@tanstack/react-router';
export function App() {
  return (
    <RuntimeConfigProvider>
      <CognitoAuth>
        <RuntimeConfigProvider>
          <RouterProvider router={router} />
        </RuntimeConfigProvider>
      </CognitoAuth>
    </RuntimeConfigProvider>
  );
}
"
`;

exports[`cognito-auth generator > should update shared constructs index.ts > common/constructs-index 1`] = `
"export * from './user-identity.js';
export * from './runtime-config.js';
"
`;
